packages/cli/src/__tests__/credentials-overwrites.test.ts

```diff
@@ -0,0 +1,134 @@
+	const parentCredentialType = mock<ICredentialType>({ name: 'parent', extends: undefined });
+
+	const globalConfig = mock<GlobalConfig>({ credentials: { overwrite: {} } });
+	const credentialTypes = mock<CredentialTypes>();
+	const logger = mock<Logger>();
+	let credentialsOverwrites: CredentialsOverwrites;
+
+	beforeEach(() => {
+		jest.resetAllMocks();
+
+		globalConfig.credentials.overwrite.data = JSON.stringify({
+			test: { username: 'user' },
+			parent: { password: 'pass' },
+		});
+		credentialTypes.recognizes.mockReturnValue(true);
+		credentialTypes.getByName.mockImplementation((credentialType) => {
+			if (credentialType === testCredentialType.name) return testCredentialType;
+			if (credentialType === parentCredentialType.name) return parentCredentialType;
+			throw new UnrecognizedCredentialTypeError(credentialType);
+		});
+		credentialTypes.getParentTypes
+			.calledWith(testCredentialType.name)
+			.mockReturnValue([parentCredentialType.name]);
+
+		credentialsOverwrites = new CredentialsOverwrites(globalConfig, credentialTypes, logger);
+	});
+
+	it('should parse and set overwrite data from config in the constructor', () => {
+		expect(credentialsOverwrites.getAll()).toEqual({
+			parent: { password: 'pass' },
+			test: {
+				password: 'pass',
+				username: 'user',
+			},
+		});
+	});
+
+	describe('setData', () => {
+		it('should reset resolvedTypes when setting new data', () => {
+			const newData = { test: { token: 'test-token' } };
+			credentialsOverwrites.setData(newData);
+
+			expect(credentialsOverwrites.getAll()).toEqual(newData);
+		});
+	});
+
+	describe('applyOverwrite', () => {
+		it('should apply overwrites only for empty fields', () => {
+			const result = credentialsOverwrites.applyOverwrite('test', {
+				username: 'existingUser',
+				password: '',
+			});
+
+			expect(result).toEqual({
+				username: 'existingUser',
+				password: 'pass',
+			});
+		});
+
+		it('should return original data if no overwrites exist', () => {
+			const data = {
+				username: 'user1',
+				password: 'pass1',
+			};
+
+			credentialTypes.getParentTypes.mockReturnValueOnce([]);
+
+			const result = credentialsOverwrites.applyOverwrite('unknownCredential', data);
+			expect(result).toEqual(data);
+		});
+	});
+
+	describe('getOverwrites', () => {
+		it('should return undefined for unrecognized credential type', () => {
+			credentialTypes.recognizes.mockReturnValue(false);
+
+			const result = credentialsOverwrites.getOverwrites('unknownType');
+
+			expect(result).toBeUndefined();
+			expect(logger.warn).toHaveBeenCalledWith(expect.stringContaining('Unknown credential type'));
+		});
+
+		it('should cache resolved types', () => {
+			credentialsOverwrites.getOverwrites('parent');
+			const firstCall = credentialsOverwrites.getOverwrites('test');
+			const secondCall = credentialsOverwrites.getOverwrites('test');
+
+			expect(firstCall).toEqual(secondCall);
+			expect(credentialTypes.getByName).toHaveBeenCalledTimes(2);
```

Comment:
What's the caching that happens here? Should this be called once instead?

Reply:
I'm not clear why we do this, but this was implemented [here](https://github.com/n8n-io/n8n/commit/6d8e9b1cd2f8f21e630641c336cf11015dc7207b).
It's being called twice here because the first call is via line 94, second via line 95, and then line 96 the cache is used.